### PR TITLE
Bring changes in main M2 wdl into unsupported m2_opt wdl

### PR DESCRIPTION
**Pull request by davidbenjamin**
@jsotobroad Would you mind reviewing this?  Earlier I brought aspects of your mutect2_opt wdl into the public m2 wdl.  Now I am bringing all the changes from the public wdl into the opt wdl while keeping your NIO-dependent code.

### CODE REVIEW

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 1]
```diff
@@ -1,767 +1,867 @@
...
+                compress = compress,
+                gatk_docker = gatk_docker,
+                preemptible_attempts = preemptible_attempts,
+                pre_adapter_metrics = input_artifact_metrics,
+                artifact_modes = artifact_modes,
+                disk_space = ceil(size(Filter.filtered_vcf, "GB") * small_input_to_output_multiplier) + ceil(size(input_artifact_metrics, "GB")) + disk_pad,
+                auth = auth
+        }
+    }
+
```
Comment (jsotobroad) [inline scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: `\n`

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 2]
```diff
@@ -1,767 +1,867 @@
...
+    String gatk_docker
+    Int? mem
+    Int? preemptible_attempts
+    Int? disk_space
+    Int? cpu
+    Boolean use_ssd = false
+
+    # Mem is in units of GB but our command and memory runtime values are in MB
+    Int machine_mem = if defined(mem) then mem * 1000 else 7000
+    Int command_mem = machine_mem - 1
```
Comment (jsotobroad) [inline scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: I think you want `- 1000` here

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 3]
```diff
@@ -1,767 +1,867 @@
...
+    command {
+        set -e
+        export GATK_LOCAL_JAR=${default="/root/gatk.jar" gatk_override}
+        gatk --java-options "-Xmx${command_mem}m" CollectSequencingArtifactMetrics \
+            -I ${tumor_bam} -O "gatk" -R ${ref_fasta} -VALIDATION_STRINGENCY LENIENT
+    }
+
+    runtime {
+        docker: gatk_docker
+        memory: machine_mem + " GB"
```
Comment (jsotobroad) [inline scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: `MB` instead of `GB`

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 4]
```diff
@@ -1,767 +1,867 @@
...
+
+        gatk --java-options "-Xmx${command_mem}m" FilterMutectCalls -V ${unfiltered_vcf} \
+      	    -O ${output_vcf} \
+      	    ${"--contamination-table " + contamination_table} \
+      	    ${m2_extra_filtering_args}
+    }
+
+    runtime {
+        docker: gatk_docker
+        memory: machine_mem + " GB"
```
Comment (jsotobroad) [inline scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: `MB` instead of `GB`

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 5]
```diff
@@ -1,767 +1,867 @@
...
+            ${"--default_config " + default_config_file}
     >>>
 
     runtime {
-      docker: oncotator_docker
-      memory: machine_mem + " MB"
-      disks: "local-disk " + disk_space_gb + " HDD"
-      preemptible: select_first([preemptible_attempts, 10])
+        docker: oncotator_docker
+        memory: machine_mem + " GB"
```
Comment (jsotobroad) [inline scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: `MB` instead of `GB`

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 6]
```diff
@@ -1,767 +1,867 @@
...
+        bootDiskSizeGb: 12
```
Comment (jsotobroad) [inline scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: whats the extra 2 GB of boot disk for?

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 7]
```diff
@@ -1,767 +1,867 @@
...
+    File? variants_for_contamination_index
+    Boolean? run_orientation_bias_filter
+    Boolean run_ob_filter = select_first([run_orientation_bias_filter, false])
+    Array[String]? artifact_modes
+    File? tumor_sequencing_artifact_metrics
+    String? m2_extra_args
+    String? m2_extra_filtering_args
+    String? split_intervals_extra_args
+    Boolean? make_bamout
+    Boolean generate_bamout = select_first([make_bamout, false])
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: Done.

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 8]
```diff
@@ -1,767 +1,867 @@
...
+    Float large_input_to_output_multiplier = 2.25
+    Float small_input_to_output_multiplier = 2.0
+
+    # logic about output file names -- these are the names *without* .vcf extensions
+    String output_basename = basename(tumor_bam, ".bam")
+    String unfiltered_name = output_basename + "-unfiltered"
+    String filtered_name = output_basename + "-filtered"
+    String funcotated_name = output_basename + "-funcotated"
+
+
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: done

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl (hunk 1 above)
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: done

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 9]
```diff
@@ -1,767 +1,867 @@
...
+    Int? mem
+    Int? preemptible_attempts
+    Int? disk_space
+    Int? cpu
+    Boolean use_ssd = false
+
+    # Mem is in units of GB but our command and memory runtime values are in MB
+    Int machine_mem = if defined(mem) then mem * 1000 else 3500
+    Int command_mem = machine_mem - 500
+
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: done

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 10]
```diff
@@ -1,767 +1,867 @@
...
-    File output_vcf_index = "${output_vcf_name}.vcf.idx"
-  }
+    # inputs
+    Array[File] input_vcfs
+    String output_name
+    Boolean compress
+    String output_vcf = output_name + if compress then ".vcf.gz" else ".vcf"
+    String output_vcf_index = output_vcf + if compress then ".tbi" else ".idx"
+
+
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: done

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl (hunk 2 above)
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: fixed in 3 places

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl (hunk 3 above)
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: fixed

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl (hunk 4 above)
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: fixed

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl [hunk 11]
```diff
@@ -1,767 +1,867 @@
...
+    Int? disk_space
+    Int? cpu
+    Boolean use_ssd = false
+
+    # Mem is in units of GB but our command and memory runtime values are in MB
+    Int machine_mem = if defined(mem) then mem * 1000 else 3500
+    Int command_mem = machine_mem - 500
+
+    command <<<
+
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: done

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl (hunk 5 above)
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: fixed

File: scripts/mutect2_wdl/unsupported/mutect2_opt.wdl (hunk 6 above)
Comment (davidbenjamin) [reply scripts/mutect2_wdl/unsupported/mutect2_opt.wdl]: No idea.  Copied from @LeeTL1220's code in mutect2.wdl.

**Comment by davidbenjamin** [comment]
@jsotobroad Back to you, assuming tests pass.  I hope @LeeTL1220 can enlighten us about `bootDiskSizeGb: 12` in the oncotator task.  In the last commit I accounted for the size of the normal bam that is now an input to the `CalculateContamination` task.

**Comment by jsotobroad** [comment]
You are going to want to change all the 
`memory: machine_mem + " GB"` 
lines to 
`memory: machine_mem + " MB"`.
Otherwise you are going to be asking for massive amounts of memory 😄.  Sorry if I wasn't clear the first time.

**Comment by davidbenjamin** [comment]
@jsotobroad You were perfectly clear.  I must have made that change on a different branch by mistake!!

**Pull request merged**